DataGrid Integration

jmacato opened the pull request:
- What does the pull request do?
Integrates https://github.com/AvaloniaUI/DataGrid to main repository for simplicity sake.
- What is the current behavior?
DataGrid is separately maintained and has to be manually added.
- What is the updated/expected behavior with this PR?
- How was the solution implemented (if it's not obvious)?

Checklist:

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

sdoroff committed:
Initial commit

sdoroff committed:
Cleaned up sample

sdoroff committed:
Updated README

sdoroff committed:
Added NuGet.Config

sdoroff committed:
Implemented Validation

Uses the binding validation framework to validate cell edits

sdoroff committed:
Merge pull request #1 from AvaloniaUI/Validation

Implemented Validation

sdoroff committed:
Added Visual Indicators for Column Reorder

sdoroff committed:
Merge pull request #3 from AvaloniaUI/column-reorder-ui

Added Visual Indicators for Column Reorder

sdoroff committed:
Fixes scrolling bug in grouped DataGrids

Fixes #5

sdoroff committed:
Merge pull request #11 from AvaloniaUI/scroll-bug

Fixes scrolling bug in grouped DataGrids

sdoroff committed:
Changed Theme Namespace to avoid conflict

sdoroff committed:
Merge pull request #12 from AvaloniaUI/change-theme-namespace

Changed Theme Namespace to avoid conflict

sdoroff committed:
Updated the Default Theme Control Templates

sdoroff committed:
Merge pull request #13 from AvaloniaUI/default-theme-changes

Updated the Default Theme Control Templates

Commit:
allow selected row background color to be overriden.

Commit:
fix crash on app load due to inconsistent namespace.

sdoroff committed:
Merge pull request #14 from danwalmsley/minor-fixes

Minor fixes

sdoroff committed:
Alter DataGridColumn to use IControl instead of Control

sdoroff committed:
Implement DataGridTemplateColumn

sdoroff committed:
Implemented RefreshCellContent on DataGridTemplateColumn

sdoroff committed:
Merge pull request #17 from AvaloniaUI/template-column

Added DataGridTemplateColumn

Commit:
Add CellPointerPressed event + Make avalonia reference nightly-invariant

sdoroff committed:
Merge pull request #18 from jmacato/cellpressedevent

Add CellPointerPressed event + Make avalonia reference nightly-invariant

kekekeks committed:
Removed root directory contents in preparation for merge with the main repo

kekekeks committed:
Merge branch 'datagrid-preparation' into datagrid-integration

Commit:
Integrate DataGrid theme to default themes project.

Commit:
DataGrid integration to Avalonia.Controls part 1 of N

Commit:
Fix namespaces, part 2 of N

Commit:
Add reference to Avalonia.Markup for binding stuff.

Commit:
Fix default theme - part 3 of N

Commit:
Integrate DataGridSample into ControlCatalog

Commit:
Add description on DataGridPage.

Commit:
Cleanup countries list.

Commit:
Merge branch 'master' into datagrid-integration

grokys commented:
I think it might be better to put this in its own `Avalonia.Control.DataGrid` assembly, as it's pretty large and I think we might want to package it independently from the core controls.

donandren commented:
if we use Datagrid in DevTools then Datagrid better be in the same Avalonia nuget package,
but in it's own assembly sound good idea, code is large and probably a subject of many changes

kekekeks commented on src/Avalonia.Controls/DataGrid/DataGridBoundColumn.cs:
```diff
@@ -0,0 +1,313 @@
+            return null;
+        }
+        
+        /*
```
This file has several commented blocks. Is it intended?

jmacato replied:
Resolved with commit 1d9593a

kekekeks commented on src/Avalonia.Controls/DataGrid/DataGridCell.cs:
```diff
@@ -0,0 +1,335 @@
+
+        internal void UpdatePseudoClasses()
+        {
+            /*
```
Same here

jmacato replied:
Also resolved with commit 1d9593a

kekekeks commented:
I see that code has support for row grouping. It would be really nice to include it to ControlCatalog sample. That will be literally the first question people will ask when they see the datagrid page

kekekeks commented:
Yep, original sample has grouping support. Row editing is also removed from ControlCatalog sample for some reason

jmacato commented:
@kekekeks all the examples on the sample project are included but the tabheaders are not displaying properly. I talked about it in the core room, i can't figure out why the inner tabcontrol inherits the style of the root sidebar tabcontrol :(

kekekeks commented:
@grokys we need your help with tabcontrol styling here

grokys commented:
@jmacato could you put the tabheaders back in and i'll do some spelunking to try to work out what's going wrong.

jmacato commented on samples/ControlCatalog/Pages/DataGridPage.xaml:
```diff
@@ -0,0 +1,54 @@
+      <TextBlock Classes="h1">DataGrid</TextBlock>
+      <TextBlock Classes="h2">A control for displaying and interacting with a data source.</TextBlock>
+    </StackPanel>
+    <TabControl Grid.Row="1">
```
@grokys the examples are already here sir

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Merge branch 'master' into datagrid-integration

jmacato commented:
@AvaloniaUI/core I'm going to do this PR again, just to double-check, we'd have to split datagrid into its own assembly & project right?

grokys commented:
@jmacato yep!

Commit:
Move DataGrid to a separate assembly and add it to CoreLibraries.props.

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Add xmlns definitions, fix references & trim some warnings.

Commit:
Temporarily fix datagrid loading by force-loading the DataGrid assembly on Markup.

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Make ICollectionView internal and rename DataGrid specific
classes on Avalonia.Collection namespace of DataGrid's assembly
with DataGrid prefix to avoid API pollution later on.

Commit:
Remove commented-out code on DataGrid.

Commit:
Remove regions + superflous empty newlines.

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Remove worthless QualityBand xml comment parameters.

grokys committed:
Merge branch 'master' into datagrid-integration

grokys commented on src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs:
```diff
@@ -0,0 +1,4315 @@
+    }
+
+    /// <summary>Defines a method that enables a collection to provide a custom view for specialized sorting, filtering, grouping, and currency.</summary>
+    internal interface ICollectionViewFactory
```
Should this be `IDataGridCollectionViewFactory`?

grokys commented on src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs:
```diff
@@ -0,0 +1,4315 @@
+    }
+
+    /// <summary>
+    /// PagedCollectionView view over an IEnumerable.
```
`PagedCollectionView` is referred to a few times in comments but doesn't seem to appear in the source.

grokys commented on src/Avalonia.Controls.DataGrid/Collections/DataGridCollectionView.cs:
```diff
@@ -0,0 +1,4315 @@
+    /// <summary>
+    /// PagedCollectionView view over an IEnumerable.
+    /// </summary>
+    public sealed class CollectionViewBase : IDataGridCollectionView, IDataGridEditableCollectionView, INotifyPropertyChanged //IPagedCollectionView, 
```
Why is this named `CollectionViewBase` when it's sealed? `Base` suggests it should be the base class for something.

Also there's commented out code with `IPagedCollectionView`.

grokys commented on src/Avalonia.Themes.Default/DataGrid.xaml:
```diff
@@ -0,0 +1,239 @@
+    </Setter>
+  </Style>
+
+  <!--FIX-->
```
What does `FIX` here mean?

grokys commented on src/Avalonia.Themes.Default/Avalonia.Themes.Default.csproj:
```diff
@@ -7,6 +7,7 @@
     <ProjectReference Include="..\Avalonia.Animation\Avalonia.Animation.csproj" />
     <ProjectReference Include="..\Avalonia.Base\Avalonia.Base.csproj" />
     <ProjectReference Include="..\Avalonia.Controls\Avalonia.Controls.csproj" />
+    <ProjectReference Include="..\Avalonia.Controls.DataGrid\Avalonia.Controls.DataGrid.csproj" />
```
Adding this reference here will cause all projects to reference `Avalonia.Controls.DataGrid`, which is what putting it in its own assembly was intended to avoid...

It would probably be better to extract the default theme from this assembly and put it in the `Avalonia.Controls.DataGrid` assembly.

grokys commented on src/Markup/Avalonia.Markup.Xaml/PortableXaml/AvaloniaRuntimeTypeProvider.cs:
```diff
@@ -38,6 +38,7 @@ public class AvaloniaRuntimeTypeProvider : IRuntimeTypeProvider
             typeof(DataTemplate).GetTypeInfo().Assembly,
             typeof(SolidColorBrush).GetTypeInfo().Assembly,
             typeof(Binding).GetTypeInfo().Assembly,
+            typeof(DataGrid).GetTypeInfo().Assembly
```
Again, this means that all avalonia applications will load the `DataGrid` assembly.

grokys commented on src/Markup/Avalonia.Markup.Xaml/Avalonia.Markup.Xaml.csproj:
```diff
@@ -58,6 +58,7 @@
     <ProjectReference Include="..\..\Avalonia.Animation\Avalonia.Animation.csproj" />
     <ProjectReference Include="..\..\Avalonia.Base\Avalonia.Base.csproj" />
     <ProjectReference Include="..\..\Avalonia.Controls\Avalonia.Controls.csproj" />
+    <ProjectReference Include="..\..\Avalonia.Controls.DataGrid\Avalonia.Controls.DataGrid.csproj" />
```
Need to remove this too.

grokys reviewed (changes requested):
Looking good, but a few changes needed. In general I think we need to treat this control like a 3rd party control, and so ship the default theme in the `Avalonia.Controls.DataGrid` assembly itself or create a separate `Avalonia.Themes.Default.DataGrid` (naming TBD) assembly.

There's also the fact that Portable.Xaml isn't properly loading all assemblies, necessitating the reference in `AvaloniaRuntimeTypeProvider`. I think for the moment we need to push the requirement to load the `Avalonia.Controls.DataGrid` assembly before XAML parsing into application code. The XAML compiler should fix this in future.

Commit:
Merge branch 'master' into datagrid-integration

Commit:
Add prefixes and correct references from PageCollectionView to DataGridCollectionView.

Commit:
Remove temporary hacks from 8ea2234.
Add a dummy function to load the DataGrid assembly before
 initializing Avalonia.
Move DataGrid theme back to Avalonia.Controls.DataGrid.

grokys committed:
Move DataGrid theme to Themes directory.

grokys committed:
Remove unnecessary reference.

grokys committed:
Merge branch 'master' into datagrid-integration

grokys committed:
Make sure DataGrid is loaded in designer.

Add an `AppBuilder` extension instead of calling `DataGrid.Load` so that the `Avalonia.Controls.DataGrid` assembly gets loaded in the designer.
